### PR TITLE
consolidate split_into_n

### LINKED ISSUE #978: better split_into_n for chunkdivision
It would be nice to replace the `split_into_three` function from #976 into a similar routine that splits into `n` for any `n`.  Basically you just call `find_best_split(n, ...)`, then `find_best_split(n-1, ...)` on the remaining chunk, then `find_best_split(n-2, ...)`, and so on `n-1` times.

This would also let us handle other prime factors (e.g. 5) more efficiently.

### PR DESCRIPTION
**Pull request by stevengj**
Closes #978.

### CODE REVIEW

File: src/vec.cpp [hunk 1]
```diff
@@ -1090,86 +1090,20 @@ void grid_volume::find_best_split(int desired_chunks, int &best_split_point,
...
-grid_volume grid_volume::split_by_cost(int desired_chunks, int proc_num, int best_split_point,
-                                       direction best_split_direction,
-                                       double left_effort_fraction) const {
-  if (desired_chunks == 1) return *this;
-
-  const int split_point = best_split_point;
-  const int num_in_split_dir = num_direction(best_split_direction);
-
-  const int num_low = (size_t)(left_effort_fraction * desired_chunks + 0.5);
-  // Revert to split() when cost method gives less grid points than chunks
```
Comment (stevengj) [inline src/vec.cpp]: I'm not sure how important this check is?  Not sure if we need to put a version of this into `split_into_n` or something?

Comment (ChristopherHogan) [reply src/vec.cpp]: That comes from the original `split_by_effort`, so I'm not sure how important it is either.

**Comment by stevengj** [comment]
@oskooi, would be nice to look at the chunk layouts for a few cases, with factors of 2, 3, and even 5.

**Comment by oskooi** [comment]
Results for this PR for various chunk counts for the [OLED example](http://www.simpetus.com/projects.html#meep_oled).

**chunks = 2, time for choose_chunkdivision = 3.43458 s**

chunk_layout_2

**chunks = 3, time for choose_chunkdivision = 8.99511 s**

chunk_layout_3

**chunks = 5, time for choose_chunkdivision = 17.7584 s**

chunk_layout_5

**chunks = 15, time for choose_chunkdivision = 84.1917 s**

chunk_layout_15

**Comment by oskooi** [comment]
Results for *prior* to this PR.

**chunks = 15, time for choose_chunkdivision = 86.3708 s**

master_chunk_layout_15

This is essentially identical to this PR.

**Comment by stevengj** [comment]
Currently the `get_prime_factors` routine does not look for prime factors of 5 or more, and instead uses fewer processes.  e.g. if you request 5 chunks, it will actually give you 4, and if you ask for 15 it will actually give you 12.

**Comment by oskooi** [comment]
Additional results for *prior* to this PR.

**chunks = 2, time for choose_chunkdivision = 3.42813 s**

master_chunk_layout_2

**chunks = 3, time for choose_chunkdivision = 9.04815 s**

master_chunk_layout_3

**chunks = 5, time for choose_chunkdivision = 17.9891 s**

master_chunk_layout_5

Again, results are nearly identical to this PR.

**Comment by stevengj** [comment]
Would be easier see the number of chunks in a case with no PML.

And my earlier comment was incorrect — it discards prime factors > 5, but keeps prime factors of 5.

**Comment by oskooi** [comment]
Same OLED example but without PML.

**chunks = 5, time for choose_chunkdivision = 26.3378 s**

master_chunk_layout_5

**Comment by stevengj** [comment]
Can you try just an empty cell?  It should produce roughly equal chunks in that case.

**Comment by oskooi** [comment]
Empty 3d cell.

**chunks = 5, time for choose_chunkdivision = 5.3503e-05 s**

master_chunk_layout_5

**Comment by stevengj** [comment]
Would be nice to verify that the weird chunk division for n=5 is faster than the equal chunk division for the case with DFT regions etc.

**Comment by stevengj** [comment]
That is, use the feature in #697 to save the chunk layout from the empty-cell case, and then load it in the case with DFT regions etcetera, and then compare timings to the chunk layout it normally comes up with in that case.

**Comment by oskooi** [comment]
For the OLED example shown in #988 (with no PML or symmetries), a comparison of the average seconds per step (the performance measure) for two different chunk layouts &mdash; empty cell vs. OLED cell &mdash; for `5` cores/chunks shows only a ~6% improvement.

**chunk layout: empty cell**
```
time for choose_chunkdivision = 20.5774 s
avg. seconds per time step: 0.17107951736450194
```

empty_chunk_layout_5

**chunk layout: OLED cell**
```
time for choose_chunkdivision = 31.5034 s
avg. seconds per time step: 0.16048019051551818
```

full_chunk_layout_5

**Comment by stevengj** [comment]
>  only a ~6% improvement.

Well, at least it's not worse.

**Comment by oskooi** [comment]
While the chunk layout for the actual cell yields consistently better performance than that for the empty cell, the performance gains are inconsistent as shown below for three cases of cores/chunks: `18` (3%), `15` (12%), and `10` (8%).

(1) Results for `18` cores/chunks shows a ~3% improvement.

**chunk layout: empty cell**
```
time for choose_chunkdivision = 43.2021 s
avg. seconds per time step: 0.05078411841392517
```

empty_chunk_layout_18

**chunk layout: OLED cell**
```
time for choose_chunkdivision = 43.1544 s
avg. seconds per time step: 0.04906289482116699
```

full_chunk_layout_18


(2) Results for `15` cores/chunks shows a 12% improvement.

**chunk layout: empty cell**
```
time for choose_chunkdivision = 42.2143 s
avg. seconds per time step: 0.055799429178237915
```

empty_chunk_layout_15

**chunk layout: OLED cell**
```
time for choose_chunkdivision = 42.2421 s
avg. seconds per time step: 0.04910236811637878
```

full_chunk_layout_15

(3) Results for `10` cores/chunks shows a ~8% improvement.

**chunk layout: empty cell**
```
time for choose_chunkdivision = 36.2672 s
avg. seconds per time step: 0.05920965790748596
```

empty_chunk_layout_10

**chunk layout: OLED cell**
```
time for choose_chunkdivision = 36.2957 s
avg. seconds per time step: 0.05466895914077759
```

full_chunk_layout_10

**Pull request merged**